crates/starknet-types-core/src/qm31/mod.rs

```diff
@@ -0,0 +1,277 @@
+
+        // Take M31 coordinates from both CM31.
+        let [c1, c2] = a.value();
+        let [c3, c4] = b.value();
+
+        (c1.to_raw(), c2.to_raw(), c3.to_raw(), c4.to_raw())
+    }
+
+    /// Packs the [QM31] into a [Felt].
+    ///
+    /// Stores the four M31 coordinates in the first 144 bits of a Felt. Each
+    /// coordinate takes 36 bits, and the resulting felt is equal to:
+    /// `C1 + C2 << 36 + C3 << 72 + C4 << 108`
+    ///
+    /// Why the stride between coordinates is 36 instead of 31? In Stwo, Felt
+    /// elements are stored in memory as 28 M31s, each representing 9 bits
+    /// (that representation is efficient for multiplication). 36 is the first
+    /// multiple of 9 that is greater than 31.
+    pub fn pack_into_felt(&self) -> Felt {
+        let (c1, c2, c3, c4) = self.to_coefficients();
+
+        // Pack as: c1 + c2 << 36 + c3 << 36*2 + c4 << 36*3.
+        let lo = c1 as u128 + ((c2 as u128) << 36);
+        let hi = c3 as u128 + ((c4 as u128) << 36);
+        let mut felt_bytes = [0u8; 32];
+        felt_bytes[0..9].copy_from_slice(&lo.to_le_bytes()[0..9]);
+        felt_bytes[9..18].copy_from_slice(&hi.to_le_bytes()[0..9]);
+        Felt::from_bytes_le(&felt_bytes)
+    }
+
+    /// Unpacks a [QM31] from the [Felt]
+    ///
+    /// See the method [QM31::pack_into_felt] for a detailed explanation on the
+    /// packing format.
+    pub fn unpack_from_felt(felt: &Felt) -> Result<QM31, InvalidQM31Packing> {
+        const MASK_36: u64 = (1 << 36) - 1;
+        const MASK_8: u64 = (1 << 8) - 1;
+
+        let digits = felt.to_le_digits();
+
+        // The QM31 is packed in the first 144 bits,
+        // the remaining bits must be zero.
+        if digits[3] != 0 || digits[2] >= 1 << 16 {
+            return Err(InvalidQM31Packing(*felt));
+        }
+
+        // Unpack as: c1 + c2 << 36 + c3 << 36*2 + c4 << 36*3.
+        let c1 = digits[0] & MASK_36;
+        let c2 = (digits[0] >> 36) + ((digits[1] & MASK_8) << 28);
+        let c3 = (digits[1] >> 8) & MASK_36;
+        let c4 = (digits[1] >> 44) + (digits[2] << 20);
+
+        // Even though we use 36 bits for each coordinate,
+        // the maximum value is still the field prime.
+        for c in [c1, c2, c3, c4] {
+            if c >= MERSENNE_31_PRIME_FIELD_ORDER as u64 {
+                return Err(InvalidQM31Packing(*felt));
+            }
+        }
+
+        Ok(QM31(Degree4ExtensionField::const_from_coefficients(
+            c1 as u32, c2 as u32, c3 as u32, c4 as u32,
+        )))
+    }
+
+    /// Multiplicative inverse inside field.
+    pub fn inverse(&self) -> Result<Self, FieldError> {
+        Ok(Self(self.0.inv()?))
+    }
+}
+
+impl Zero for QM31 {
+    fn zero() -> Self {
+        Self(FieldElement::<Degree4ExtensionField>::zero())
+    }
+
+    fn is_zero(&self) -> bool {
+        self == &Self::zero()
+    }
+}
+impl One for QM31 {
+    fn one() -> Self {
+        Self(FieldElement::<Degree4ExtensionField>::one())
+    }
+}
```

Comment:
Implementation of traits comming from `num-traits` crate should be behind a feature flag.
Check the way we did it for Felt in `crates/starknet-types-core/src/felt/num_traits_impl.rs`